ingestion/src/lib.rs

```diff
@@ -170,6 +269,67 @@ pub async fn insert_data(data: Data<'_>, conn: &mut PooledPgConn<'_>) -> Result<
     Ok(())
 }
 
+pub async fn qc_data(
+    chunks: &mut Vec<DataChunk<'_>>,
+    rove_connector: &rove_connector::Connector,
+    pipelines: &HashMap<(i32, RelativeDuration), rove::Pipeline>,
+) -> Result<Vec<QcProvenance>, Error> {
+    let mut qc_results: Vec<QcProvenance> = Vec::new();
+    for chunk in chunks {
+        let time_resolution = match chunk.time_resolution {
+            Some(time_resolution) => time_resolution,
+            // if there's no time_resolution, we can't QC
+            None => continue,
+        };
+        let timestamp = chunk.timestamp.timestamp();
+
+        for datum in chunk.data.iter_mut() {
+            let time_spec =
+                TimeSpec::new(Timestamp(timestamp), Timestamp(timestamp), time_resolution);
+            let pipeline = match pipelines.get(&(datum.param_id, time_resolution)) {
+                Some(pipeline) => pipeline,
+                None => continue,
+            };
+            let data = rove_connector
+                .fetch_one(
+                    datum.timeseries_id,
+                    &time_spec,
+                    pipeline.num_leading_required,
+                    pipeline.num_trailing_required,
+                )
+                .await?;
+            let rove_output = rove::Scheduler::schedule_tests(pipeline, data)?;
+
+            let first_fail = rove_output.iter().find(|check_result| {
+                if let Some(result) = check_result.results.first() {
+                    if let Some(flag) = result.values.first() {
+                        return *flag == rove::Flag::Fail;
+                    }
+                }
+                false
+            });
```

Comment:
Here you are using `first` because we expect a single timeseries and single data point? I would either add a comment about it or generalize it. And do we plan to use this same function to run periodic/consistency checks? If not. I'd simply rename this one to `qc_fresh_data`

Reply:
> Here you are using first because we expect a single timeseries and single data point?

Yes

> And do we plan to use this same function to run periodic/consistency checks?

No, I think their ideal signatures are too different

I'll add clarifying comments, and rename the function

Reply:
Done